### PR TITLE
Improvements in i18n support

### PR DESCRIPTION
**Pull request by iionly**

### CODE REVIEW

**Comment by juho-jaakkola** [comment]
Better approach would be to allow translating the date format itself.

``` php
$date_format = elgg_echo('event_calendar:date_format');
$date = date($date_format, $event->start_time);
```

That way it would be possible to use different date format depending on the language.

**Comment by juho-jaakkola** [comment]
If there is a need for multiple contexts, we could use:
- `event_calendar:date_format:short` e.g. 25.3.
- `event_calendar:date_format:full` e.g. 25 March 2015
- `event_calendar:date_format:foo_bar`

**Comment by juho-jaakkola** [comment]
I would also suggest replacing this:

``` php
$date = elgg_echo('foo', $foo) . ': ' . elgg_echo('bar', $bar);
```

With:

``` php
$date = elgg_echo('foo_bar', array($foo, $bar);

// Language file:
array(
    'foo_bar' => 'Foo %s: bar %s',
)
```

We cannot expect that all possible languages want to use colon in that exact position. Or that they want to use colon at all.

**Comment by iionly** [comment]
How would it be possible to get the translations of the month names into

``` php
$date_format = elgg_echo('event_calendar:date_format');
$date = date($date_format, $event->start_time);
```

?
date() has a parameter to get the full month names ('F') but the output is English only - this was the way it was implemented in event_calendar before. I don't know how to get the date displayed in the user's language if the format would be directly defined by a language string.

**Comment by juho-jaakkola** [comment]
Oops, it should in fact be [`strftime()`](http://php.net/manual/en/function.strftime.php) instead of `date()`. Then you can use [`setlocale()`](http://php.net/manual/en/function.setlocale.php) to define the language used in the weekday and month names.

**Comment by iionly** [comment]
Have you ever used setlocale() on an Elgg project? Does it work in a reliable way? I tried using it in my Celebrations plugin but then given up because I couldn't get it working as I wanted it to work.

I see several problems (maybe these are no real problems...):
- setlocale() changes locale information per process (not per thread). I wouldn't want to interfere with other scripts that might rely on a certain locale setting.
- how to be sure that all required locale information is available on any given server?
- locale language codes are different from Elgg language codes. The code retrieved from a user's language setting (fallback to site language an second fallback to English) would have to be converted into the locale language code (which could have an additional variety depending on OS of server).

**Comment by iionly** [comment]
@juho-jaakkola I've updated the commit. The colon is now part of the corresponding (new) language strings to allow for modifying / omitting.

I've not changed the code to use setlocale() though - simply because I don't know how to implement it in any useful way.

**Comment by juho-jaakkola** [comment]
I don't have experience on using setlocale() either. But I think it would be a better overall solution in case it works reliably. I'll do some tests with it to decide the best approach.

**Comment by juho-jaakkola** [comment]
My colleague said that that Moodle is using the server's locale. I'll take a look how they've implemented it.

**Comment by juho-jaakkola** [comment]
Moodle has quite a lot of different formats: https://github.com/moodle/moodle/blob/master/lang/en/langconfig.php

File: languages/en.php [hunk 1]
```diff
@@ -34,6 +34,9 @@
 	'event_calendar:day_label' => "Day",
 	'event_calendar:week_label' => "Week",
 	'event_calendar:month_label' => "Month",
+	'event_calendar:day_label_with_separator' => "Day: ",
+	'event_calendar:week_label_with_separator' => "Week: ",
+	'event_calendar:month_label_with_separator' => "Month: ",
```
Comment (juho-jaakkola) [inline languages/en.php]: The date string should also be included in the translation:

``` php
'event_calendar:month_label_with_separator' => "Month: %s",
```

There are languages that are read from right to left, so then the translation might be:

``` php
'event_calendar:month_label_with_separator' => "%s شهر",
```

(Just an example. I don't know the correct Arabic format.)

Comment (iionly) [inline languages/en.php]: I guess this is a valid use case indeed. I'll look into it. Though I won't be able to work on it this week anymore.

**Comment by iionly** [comment]
There might be some differences to the approach of Moodle to be worth taking into account:
- It looks to me that the Moodle core engine already provides support for different time/date formats (I don't know Moodle and therefore I don't know if it's correct to refer to Moodle core). If Elgg core would offer the usage of different formats already, it would surely be the better way compared to implementing this on a per-plugin basis. I believe there are also some open issues for Elgg core at least on improval of timezone support. Maybe it would be a good idea to think about adding date/time format support in core together with improving / adding timezone support.
- It looks to me that Moodle doesn't use setLocale of php itself but rather builts on the Zend Framework. Setting locale via Zend might not have the problem of php's setLocale being a global setting.
- The language file of Moodle makes me guess that they are not relying on some built-in mapping of dfferent language code standards. It looks to me that the different codes are provided as strings instead. This could be a possibility for event_calendar, too (though it looks like an awful lot of possibilities to consider:

``` php
$string['iso6391'] = 'en';
$string['iso6392'] = 'eng';
$string['labelsep'] = ': ';
$string['listsep'] = ',';
$string['locale'] = 'en_AU.UTF-8';
$string['localewin'] = 'English_Australia.1252';
$string['localewincharset'] = '';
$string['oldcharset'] = 'ISO-8859-1';
```

(especially if you might even have different OS on server and/or client side into account).

**Comment by iionly** [comment]
@juho-jaakkola PR updated.

**Comment by juho-jaakkola** [comment]
I tried `setlocale()` and it worked perfectly. Also it took only a couple of seconds to add Finnish as a new locale to my server. I'm now looking into the issues you listed above.

**Comment by juho-jaakkola** [comment]
There are also javascript libraries that support the strftime syntax: https://github.com/samsonjs/strftime

**Comment by iionly** [comment]
I'm still not really sure about usage of setlocale(). On the one hand, it's of course not difficult to implement it within the plugin. On the other hand, I still think it can get complicated for users to configure it (different encodings) and likely even impossible in some cases (I don't think you would be able to add missing locales on a shared server). Additionally, I'm still not sure if setlocale() might cause conflicts with other scripts (after reading the warning on the php docs page and also some comments about it).

JS might be a possibility (maybe using jquere datepicker.parseDate might work, too).

Would it be possible to decide on this separately? The problem is that any change I would like to make would involve modifications in model.php. And I really don't want to constantly rebase if I make these PRs in parallel.

Another possibility could be to not implement much more in terms of i18n for now but wait for Elgg 2.0 (which I'm sure will require some rewrites anyway). It looks like Elgg 2.0 might bring some changes / improvements in terms of i18n and language strings support. If these changes get implemented it would be much easier to realize different date formats.

**Comment by juho-jaakkola** [comment]
> Would it be possible to decide on this separately?

Sure.

**Pull request merged**